convert2rhel/actions/__init__.py

```diff
@@ -0,0 +1,619 @@
+                # Uncaught exceptions are handled by constructing a generic
+                # failure message here that should be reported
+                message = (
+                    "Unhandled exception was caught: %s\n"
+                    "Please file a bug at https://issues.redhat.com/ to have this"
+                    " fixed or a specific error message added.\n"
+                    "Traceback: %s" % (e, traceback.format_exc())
+                )
+                action.set_result(status="ERROR", error_id="UNEXPECTED_ERROR", message=message)
+
+            # Categorize the results
+            if action.status <= STATUS_CODE["WARNING"]:
+                logger.info("%s has succeeded" % action.id)
+                successes.append(action)
+
+            if action.status > STATUS_CODE["WARNING"]:
+                message = format_report_message(action.status, action.id, action.error_id, action.message)
+                logger.error(message)
+                failures.append(action)
+                failed_action_ids.add(action.id)
+
+        if self.next_stage:
+            successes, failures, skips = self.next_stage.run(successes, failures, skips)
+
+        return FinishedActions(successes, failures, skips)
+
+
+def resolve_action_order(potential_actions, previously_resolved_actions=None):
+    """
+    Order the Actions according to the order in which they need to run.
+
+    :param potential_actions: Sequence of Actions which we need to find the
+        order of.
+    :type potential_actions: Sequence
+    :param previously_resolved_actions: Sequence of Actions which have already
+        been resolved into dependency order.
+    :type previously_resolved_actions: Sequence
+    :raises DependencyError: when it is impossible to satisfy a dependency in
+        an Action.
+    :returns: Iterator of Actions sorted so that all dependent Actions are run
+        before actions which depend on them.
+
+    .. note::
+
+        * The sort is stable but not predictable. The order will be the same
+          as long as the Actions given has not changed but adding or
+          subtracting Actions or dependencies can alter the order a lot.
+
+        * The returned actions do not include ``previously_resolved_actions``.
+          It is up to the caller to processs those before the actions returned
+          by this function.
+    """
+    if previously_resolved_actions is None:
+        previously_resolved_actions = []
+
+    # Sort the potential actions before processing so that the dependency
+    # order is stable. (Always yields the same order if the input and
+    # algorithm has not changed)
+    potential_actions = sorted(potential_actions, key=lambda action: action.id)
+
+    # actions which have yet to be sorted
+    unresolved_actions = []
+    # ids of the actions which have already been sorted
+    resolved_action_ids = set(action.id for action in previously_resolved_actions)
```

Comment:
Here it shouldn't be something like: 

```suggestion
    # actions which have yet to be resolved
    unresolved_actions = []
    # ids of the actions which have already been resolved
    resolved_action_ids = set(action.id for action in previously_resolved_actions)
```

Reply:
I agree, resolved is the better word here.  Resolved really means: Action has been sorted into its final order and yielded to the user.  So sorted is a subset of the steps that are meant.  (Additionally, we performed a pre-sort of the Actions a few lines earlier so using "sorted" again here could be confused with that.)